tests/generator.py

```diff
@@ -0,0 +1,480 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import io
+import logging
+
+from pymeasure.adapters import VISAAdapter
+from pymeasure.instruments import Channel
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+def write_generic_test(file, header_text, cls_name, comm_text, test, inkwargs=None):
+    """Write a generic test.
+
+    :param file: File to write to.
+    :param list header_text: Text of the header (parametrization, test name etc.)
+    :param cls_name: Name of the instrument class.
+    :param list comm_pairs: List of communication pairs.
+    :param str test: Test to assert for.
+    :param dict inkwargs: Dictionary of instrument instantiation kwargs.
+    """
+    if inkwargs is None:
+        args_text = "",
+    else:
+        args_text = [f'            {key}={repr(value)},\n' for key, value in inkwargs.items()]
+    inst = " as inst" if "inst" in test else ""
+    file.writelines([
+        "\n",
```

Comment:
Exactly the list expansion creates difficulties. An option would be:
~~~python
    file.write(
        f"""

{''.join(header_text)}    with expected_protocol(
            {cls_name},
{''.join(comm_text)}{''.join(args_text)}    ){inst}:
        {test}
"""
    )
~~~

It does not look that much better, but a little bit.

I cannot put the curly braces in a new line, in case the iterable is empty.

Reply:
I managed to put it into extra lines with an spare `\`.